File: compatLib/src/app/lawnchair/compatlib/RecentsAnimationRunnerCompat.java
package app.lawnchair.compatlib;

import android.graphics.Rect;
import android.view.IRecentsAnimationController;
import android.view.RemoteAnimationTarget;
import android.window.TaskSnapshot;

public interface RecentsAnimationRunnerCompat {

    void onAnimationStart(
            IRecentsAnimationController controller,
            RemoteAnimationTarget[] apps,
            RemoteAnimationTarget[] wallpapers,
            Rect homeContentInsets,
            Rect minimizedHomeBounds);

    /** Called only in T platform */
    void onAnimationCanceled(int[] taskIds, TaskSnapshot[] taskSnapshots);

    /** Called only in Q/R/S platform */
    void onAnimationCanceled(Object taskSnapshot);

    /** Called only in R/S platform */
    void onTaskAppeared(RemoteAnimationTarget app);

    /** Called only in T+ platform */
    void onTasksAppeared(RemoteAnimationTarget[] apps);
}

File: compatLib/src/app/lawnchair/compatlib/RecentsAnimationRunnerCompat.kt
package app.lawnchair.compatlib;

import android.graphics.Rect
import android.view.IRecentsAnimationController
import android.view.RemoteAnimationTarget
import android.window.TaskSnapshot

interface RecentsAnimationRunnerCompat {

    fun onAnimationStart(
        controller: IRecentsAnimationController,
        apps: Array<RemoteAnimationTarget>,
        wallpapers: Array<RemoteAnimationTarget>,
        homeContentInsets: Rect,
        minimizedHomeBounds: Rect
    )

    /**
     * Called only in T platform
     */
    fun onAnimationCanceled(taskIds: IntArray, taskSnapshots: Array<TaskSnapshot>)

    /**
     * Called only in Q/R/S platform
     */
    fun onAnimationCanceled(taskSnapshot: Any)

    /**
     * Called only in R/S platform
     */
    fun onTaskAppeared(app: RemoteAnimationTarget)

    /**
     * Called only in T+ platform
     */
    fun onTasksAppeared(apps: Array<RemoteAnimationTarget>)
}
